exporters/metric/cortex/auth.go

```diff
@@ -88,3 +92,99 @@ func (e *Exporter) addBearerTokenAuth(req *http.Request) error {
 
 	return nil
 }
+
+// buildClient returns a http client that uses TLS and has the user-specified proxy and
+// timeout.
+func (e *Exporter) buildClient() (*http.Client, error) {
+	// Create a TLS Config struct for use in a custom HTTP Transport.
+	tlsConfig, err := e.buildTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	// Convert proxy url to proxy function for use in the created Transport.
+	if e.config.ProxyURL != "" {
+		proxyURL, err := url.Parse(e.config.ProxyURL)
+		if err != nil {
+			return nil, err
+		}
+		proxy := http.ProxyURL(proxyURL)
+		transport.Proxy = proxy
+	}
+
+	client := http.Client{
+		Transport: transport,
+		Timeout:   e.config.RemoteTimeout,
+	}
+	return &client, nil
+}
+
+// buildTLSConfig creates a new TLS Config struct with the properties from the exporter's
+// Config struct.
+func (e *Exporter) buildTLSConfig() (*tls.Config, error) {
+	tlsConfig := &tls.Config{}
+	if e.config.TLSConfig == nil {
+		return tlsConfig, nil
+	}
+
+	// Set the server name if it exists.
+	if e.config.TLSConfig["server_name"] != "" {
+		tlsConfig.ServerName = e.config.TLSConfig["server_name"]
+	}
+
+	// Set InsecureSkipVerify. Viper reads the bool as a string since it is in a map.
+	parsedBool, err := strconv.ParseBool(e.config.TLSConfig["insecure_skip_verify"])
+	if err != nil {
+		return nil, err
+	}
+	tlsConfig.InsecureSkipVerify = parsedBool
+
+	// Load certificates from CA file if it exists.
+	if err := e.loadCACertificate(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	// Load the client certificate if it exists.
+	if err := e.loadClientCertificate(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return tlsConfig, nil
+}
+
+// loadCACertificates reads and loads a CA certificate file and updates a TLS Config
+// struct's certificate pool with it.
+func (e *Exporter) loadCACertificate(tlsConfig *tls.Config) error {
+	caFile := e.config.TLSConfig["ca_file"]
+
+	if caFile != "" {
+		caFileData, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return err
+		}
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM(caFileData)
+		tlsConfig.RootCAs = certPool
+	}
+	return nil
+}
+
+// loadClientCertificate loads a x509 certificate from a certificate file and key file,
+// and stores it in a TLS Config struct.
+func (e *Exporter) loadClientCertificate(tlsConfig *tls.Config) error {
```

Comment:
Is there a reason these need to be separate functions? It doesn't seem like they are called anywhere else and could just be an inline conditional. It would be my preference to do that instead and avoid mutating input on a function call.

Reply:
No reason other than making the `buildClient` method smaller. I'll change them to inline conditionals.